rhombus/private/treelist.rkt

```diff
@@ -0,0 +1,1077 @@
+
+(define empty-node (Node (vector)))
+
+(define (leaf v) (Node (vector v)))
+
+(struct treelist (root size height)
+  #:property prop:equal+hash (list
+                              (lambda (tl other-tl recur)
+                                (treelist-equal? tl other-tl recur))
+                              (lambda (v recur)
+                                (treelist-hash-code v recur))
+                              (lambda (v recur)
+                                42))
+  #:property prop:sequence (lambda (tl)
+                             (make-do-sequence
+                              (lambda ()
+                                (values
+                                  (lambda (i) (treelist-ref tl i))
+                                  add1
+                                  0
+                                  (lambda (i) (i . < . (treelist-length tl)))
+                                  #f
+                                  #f)))))
+
+(define empty-treelist (treelist empty-node 0 0))
+
+(define (treelist-empty? v)
+  (or (eq? v empty-treelist)
+      (and (impersonator? v)
+           (equal? v empty-treelist))))
+
+(define make-treelist
+  (let ([treelist
+         (case-lambda
+           [() empty-treelist]
+           [(a) (treelist (leaf a) 1 0)]
+           [(a b) (treelist (Node (vector a b)) 2 0)]
+           [(a b c) (treelist (Node (vector a b c)) 3 0)]
+           [ds (list->treelist ds)])])
+    treelist))
+
+(define (check-treelist who tl)
+  (unless (treelist? tl)
+    (raise-argument-error* who 'racket/primitive "treelist?" tl)))
+
+(define (check-treelist-index who tl size index)
+  (unless (fixnum? index)
+    (if (exact-nonnegative-integer? index)
+        (raise-argument-error* who 'racket/primitive "exact-nonnegative-integer?" index)
+        (raise-range-error* who 'racket/primitive "treelist" "" index tl 0 (fx- size 1))))
+  (when (or (index . fx< . 0) (index . fx>= . size))
+    (raise-range-error* who 'racket/primitive "treelist" "" index tl 0 (fx- size 1))))
+
+(define (check-treelist-end-index who tl size index)
+  (unless (fixnum? index)
+    (if (exact-nonnegative-integer? index)
+        (raise-argument-error* who 'racket/primitive "exact-nonnegative-integer?" index)
+        (raise-range-error* who 'racket/primitive "treelist" "" index tl 0 size)))
+  (when (or (index . fx< . 0) (index . fx> . size))
+    (raise-range-error* who 'racket/primitive "treelist" "" index tl 0 size)))
+
+(define-sequence-syntax in-treelist
+  (lambda () #'in-treelist/proc)
+  (lambda (stx)
+    (syntax-case stx ()
+      [[(d) (_ tl-expr)]
+       #'[(d)
+          (:do-in
+           ([(tl) tl-expr])
+           (unless (variable-reference-from-unsafe? (#%variable-reference))
+             (check-treelist 'in-treelist tl))
+           ([pos 0]
+            [node empty-node]
+            [node-pos 0])
+           (pos . fx< . (treelist-size tl))
```

Comment:
Do we want to record the size instead of computing it every time?

Reply:
That's probably better, but I'm not certain. The `treelist-size` function is just a field accessor, and before treelists supported chaperones, it would not be obviously slower than a closure lookup or loop-carried variable (in addition to `tl`, which must still be kept). But now that treelissts support chaperones, it's more likely faster to extract the size once.

Reply:
This does not make enough of a difference to measure reliably on my machine, but since it doesn't hurt, I'll lift out the size lookup.